// polytrix-core-api/drives/tesseract/info/index.js
var Q = require('q');
var config = require('../../config').tesseract;
var req = require('request');
var authenticator = require('../authenticator');

module.exports.accountInfo = function(access_token,refresh_token){
	var url = config.domain + '/vault/';
	return authenticator('GET', access_token, url)
        .then(transformUsageQouta)
	    .catch(transformError);
};

module.exports.usageQouta = function(access_token,refresh_token){
    var url = config.domain + '/vault/';
    return authenticator('GET', access_token, url)
        .then(transformUsageQouta)
        .catch(transformError);
};

function transformUsageQouta(raw){
    return {
        total_bytes:raw.capacity,
        used_bytes:raw.used,
        _raw:raw
    };
}

function transformError(err){
    var result = {
        success: false,
        error_code: err.status,
        msg: err.message
    };
    if(err instanceof Error){
        console.log(err.stack);
        result.error_code = 500;
        result.msg = err.message;
    }
    return result;
}
