// polytrix-core-api/drives/config.js

var domain = 'io.dev';
var port = '8000';

module.exports =

{
	dropbox : {
		client_key : 'jixx3poq3ysx5fo',
		client_secret : 'q6een1n7vuro900',
		redirect_url : 'https://'+domain+':'+port+'/api/auth/dropbox/redirect/',
	},
	onedrive : {
		client_key : '000000004413589F',
		client_secret : '9ozLAtytWthS-MROn9Z7CK7be7DJjZ3a',
		scope: 'wl.offline_access wl.skydrive_update',
		redirect_url: 'https://'+domain+':'+port+'/api/auth/onedrive/redirect/',
	},
	googledrive : {
		client_key : '145548963063-pu5kgupcpuepa3h8p29rn0ctol34sv8p.apps.googleusercontent.com',
		client_secret : 'xA7EkEq6P3TI1dhtx7b03LtF',
		scope : 'https://www.googleapis.com/auth/drive https://www.googleapis.com/auth/contacts.readonly',
		access_type : 'offline',
		redirect_url : 'https://'+domain+':'+port+'/api/auth/googledrive/redirect/',
	},
	tesseract : {
		apiKey : 'd6a13443bb2decba8e9d5fc6ffd91bc5fa04a83abdd44b3030ef9fb36a40fe08526dc893f00ae7ecaa463dc720b73352993013153b9663406f5b9b92',
		domain : 'https://io.dev:3010/api/'
	}
};
