// polytrix-core-api/drives/config.js

var domain = 'io.dev';
var port = '8000';

module.exports =

{
	dropbox : {
		client_key : 'jixx3poq3ysx5fo',
		client_secret : 'q6een1n7vuro900',
		redirect_url : 'https://'+domain+':'+port+'/api/auth/dropbox/redirect/',
	},
	onedrive : {
		client_key : '000000004413589F',
		client_secret : '9ozLAtytWthS-MROn9Z7CK7be7DJjZ3a',
		scope: 'wl.offline_access wl.skydrive_update',
		redirect_url: 'https://'+domain+':'+port+'/api/auth/onedrive/redirect/',
	},
	googledrive : {
		client_key : '145548963063-pu5kgupcpuepa3h8p29rn0ctol34sv8p.apps.googleusercontent.com',
		client_secret : 'xA7EkEq6P3TI1dhtx7b03LtF',
		scope : 'https://www.googleapis.com/auth/drive https://www.googleapis.com/auth/contacts.readonly',
		access_type : 'offline',
		redirect_url : 'https://'+domain+':'+port+'/api/auth/googledrive/redirect/',
	},
	tesseract : {
		apiKey : '31e735f3da507cb3b3c8c4909b35152d544c7595f956498b791ef56faab1a5461f929a0e337057b8574b3edf422a1d15e0f5aea4745525152b672711',
		domain : 'https://io.dev:3010/api/'
	}
};
