// polytrix-core-api/drives/dropbox.js
var filefog = require('filefog');
var config = require('./config').dropbox;

filefog.use('dropbox',require('filefog-dropbox'),{
	client_key : config.client_key,
	client_secret : config.client_secret,
	redirect_url: config.redirect_url
});

var dropboxProvider = filefog.provider('dropbox');


//
//===== public functions =====
//
var getAuthURL = function(){
	return dropboxProvider.oAuthGetAuthorizeUrl();
}

var setToken = function(req){
	return dropboxProvider.oAuthGetAccessToken(req.query.code)
	.then(function(results)
	{
		var sess = req.session;
		sess.dropbox = sess.dropbox || new Object(); //initialize dropbox object if not created.
		sess.dropbox.access_token = results.access_token;
		sess.dropbox.refresh_token = results.refresh_token;
		console.log(results);
		return sess.dropbox;
	})
	.catch(function(err){
		console.log(err);
	});
};

//	
//	@method getFileIndex
//	@params{string} location
//	@params{string} access_token
//	@params{string} refresh_token
//	@return-promise-value: fileIndex
//
var getFileIndex = function(location,access_token,refresh_token){
	refresh_token = null;//dropbox no refresh token

	console.log(location);
	console.log(access_token);

	return filefog.client("dropbox", {
		access_token: access_token,
		refresh_token: null
	})
	.then(function (client) {
		return client.retrieveFolderItems(location);
	});
};

//	
//	@method downloadFile
//	@params{string} location
//	@params{string} access_token
//	@params{string} refresh_token
//	@return-promise-value: {file,contentType,contentLength}
//
var downloadFile = function(location,access_token,refresh_token){
	refresh_token = null;//dropbox no refresh token
	
	return filefog.client("dropbox", {
		access_token: access_token,
		refresh_token: refresh_token
	})
	.then(function (client) {
		return client.downloadFile(location);
	}).then(function (data){
		console.log(data);

		var result = {
			file: data.data,
			contentType: data._raw.stat.mimeType,
			contentLength: data._raw.stat.size
		};
		return result;
	});
};

module.exports.getAuthURL = getAuthURL;
module.exports.setToken = setToken;
module.exports.getFileIndex = getFileIndex;
module.exports.downloadFile = downloadFile;
