// polytrix-core-api/drives/googledrive.js
var filefog = require('filefog');
var config = require('./config').googledrive;

filefog.use('googledrive',require('filefog-google'),{
	client_key : config.client_key,
	client_secret : config.client_secret,
	scope : config.scope,
	redirect_url: config.redirect_url
});

var googledriveProvider = filefog.provider('googledrive');


//
//===== public functions =====
//
var getAuthURL = function(){
	return googledriveProvider.oAuthGetAuthorizeUrl();
}

var setToken = function(req){
	console.log('googledrive.js setToken');
	return googledriveProvider.oAuthGetAccessToken(req.query.code)
	.then(function(results)
	{
		var sess = req.session;
		sess.googledrive = sess.googledrive || new Object(); //initialize googledrive object if not created.
		sess.googledrive.access_token = results.access_token;
		sess.googledrive.refresh_token = results.refresh_token;
		console.log(results);
		return sess.googledrive;
	})
	.catch(function(err){
		console.log(err);
	});
};

//	
//	@method getFileIndex
//	@params{string} location
//	@params{string} access_token
//	@params{string} refresh_token
//	@return-promise-value: fileIndex
//
var getFileIndex = function(location,access_token,refresh_token){
	console.log('googledrive.js getFileIndex');
	console.log('access_token:' + access_token);
	console.log('refresh_token:' + refresh_token);
	googledriveClient = filefog.client("googledrive", {
		access_token: access_token,
		refresh_token: refresh_token
	})
	.then(function (client) {
		return client.retrieveFolderItems(location);
	});
	return googledriveClient;
};

//	
//	@method downloadFile
//	@params{string} location
//	@params{string} access_token
//	@params{string} refresh_token
//	@return-promise-value: {file,contentType,contentLength}
//
var downloadFile = function(location,access_token,refresh_token){
	console.log('googledrive.js downloadFile');
	return filefog.client("googledrive", {
		access_token: access_token,
		refresh_token: refresh_token
	})
	.then(function (client) {
		return client.downloadFile(location);
	}).then(function (data){
		console.log(data.header);
		var result = {
			file: data.data,
			// contentType: data._raw.stat.mimeType,
			// contentLength: data._raw.stat.size
		};
		return result;
	});
};

module.exports.getAuthURL = getAuthURL;
module.exports.setToken = setToken;
module.exports.getFileIndex = getFileIndex;
module.exports.downloadFile = downloadFile;
