// polytrix-core-api/drives/googledrive.js
var filefog = require('filefog');
var config = require('./config').googledrive;

filefog.use('googledrive',require('filefog-google'),{
	client_key : config.client_key,
	client_secret : config.client_secret,
	scope : config.scope,
	redirect_url: config.redirect_url
});

var googledriveProvider = filefog.provider('googledrive');


//
//===== public functions =====
//
var getAuthURL = function(){
	return googledriveProvider.oAuthGetAuthorizeUrl();
}

var getToken = function(code){
	console.log('googledrive.js setToken');
	return googledriveProvider.oAuthGetAccessToken(code)
	.then(function(results)
	{
		console.log(results);
		return {
			access_token: results.access_token,
			refresh_token: results.refresh_token
		};
	})
	.catch(function(err){
		console.log(err);
		return err;
	});
};

var renewToken = function(refresh_token){
	console.log('googledrive.js renewToken');
	return googledriveProvider.oAuthRefreshAccessToken({refresh_token:refresh_token})
	.then(function(newOAuthData){
		console.log(newOAuthData);
	});

};

//	
//	@method getFileIndex
//	@params{string} location
//	@params{string} access_token
//	@params{string} refresh_token
//	@return-promise-value: fileIndex
//
var getFileIndex = function(location,access_token,refresh_token){
	console.log('googledrive.js getFileIndex');
	console.log('access_token:' + access_token);
	console.log('refresh_token:' + refresh_token);
	return googledriveClient = filefog.client("googledrive", {
		access_token: access_token,
		refresh_token: refresh_token
	})
	.then(function (client) {
		return client.retrieveFolderItems(location);
	}).then(function(result){
		result.success = true;
		return result;
	}).catch(function (err){
		console.log(err);
		var result = {
			success: false,
			msg: err.code==404 ? '404 file not found' : err.message
		}
		return result;
	});
};

//	
//	@method downloadFile
//	@params{string} location
//	@params{string} access_token
//	@params{string} refresh_token
//	@return-promise-value: {file,contentType,contentLength}
//
var downloadFile = function(location,access_token,refresh_token){
	console.log('googledrive.js downloadFile');
	return filefog.client("googledrive", {
		access_token: access_token,
		refresh_token: refresh_token
	})
	.then(function (client) {
		return client.downloadFile(location);
	}).then(function (data){
		console.log('data:');
		console.log(data);
		var result = {
			success: true,
			file: data.data,
			contentType: data.headers['content-type'],
			contentLength: data.headers['content-length'],
			contentDisposition: data.headers['content-disposition']
		};
		return result;
	}).catch(function (err){
		if(err.message == 'Invalid Credentials'){
			//
		}
		console.log(err);
		var result = {
			success: false,
			msg: err.code==404 ? '404 file not found' : err.message
		}
		return result;
	});
};

module.exports.getAuthURL = getAuthURL;
module.exports.getToken = getToken;
module.exports.getFileIndex = getFileIndex;
module.exports.downloadFile = downloadFile;
module.exports.renewToken = renewToken;
