// polytrix-core-api/drives/service.js
//
//	Service.js
//	This is the standard class and wrapper for all drive services
//

// auth
//	authUrl()
//	getToken(code)
//	renewToken(refresh_token)

// info
//	accountInfo(access_token,refresh_token)
//	usageQouta(access_token,refresh_token)

// metadata
//	get(access_token,refresh_token,fileId)
//	listFile(access_token,refresh_token,fileId)
//	downloadUrl(access_token,refresh_token,fileId)

// operation
//	create(access_token,refresh_token,destinationFileId,...(filename,etc.))
//	rename(access_token,refersh_token,fileId,...(filename,etc.))
//	move(access_token,refresh_token,fileId,destinationFileId)
//	copy(access_token,refresh_token,fileId,destinationFileId)
//	delete(access_token,refresh_token,fileId)

// stream
//	createAndUploadStream(access_token,refresh_token,destinationFileId,...(filename,etc.))
//	uploadStream(access_token,refresh_token,fileId,...(patches?))
//	downloadStream(access_token,refresh_token,fileId)

// share
//	list(access_token,refresh_token,fileId)
//	get(access_token,refresh_token,fileId)
//	new(access_token,refresh_token,fileId,...)
//	update(access_token,refresh_token,fileId,...)
//	remove(access_token,refresh_token,fileId)
var Client = require('./client');


var Service = function(serviceProvider){
	this.provider = serviceProvider;
	this.auth.provider = serviceProvider;
	this.info.provider = serviceProvider;
	this.operation.provider = serviceProvider;
	this.stream.provider = serviceProvider;
	this.share.provider = serviceProvider;
	this.rootIdentifier = this.provider.rootIdentifier;
};

// auth

Service.prototype.auth = {
	authUrl: function(){
		return this.provider.authUrl();
	},

	getToken: function(code){
		return this.provider.getToken(code);
	},

	renewToken: function(refresh_token){
		return this.provider.renewToken(refresh_token);
	},
};



// info

Service.prototype.info = {
	accountInfo: function(access_token,refresh_token){
		return this.provider.accountInfo(access_token,refresh_token);
	},

	usageQouta: function(access_token,refresh_token){
		return this.provider.usageQouta(access_token,refresh_token);
	},
};


// metadata

Service.prototype.metadata = {
	get: function(access_token,refresh_token,fileId){
		return this.provider.get(access_token,refresh_token,fileId);
	},

	listFile: function(access_token,refresh_token,fileId){
		return this.provider.listFile(access_token,refresh_token,fileId);
	},

	downloadUrl: function(access_token,refresh_token,fileId){
		return this.provider.downloadUrl(access_token,refresh_token,fileId);
	},
};


// operation

Service.prototype.operation = {
	create: function(access_token,refresh_token,destinationFileId,    others){
		return this.provider.create(access_token,refresh_token,destinationFileId,    others);
	},

	rename: function(access_token,refresh_token,fileId,    others){
		return this.provider.rename(access_token,refresh_token,fileId,    others);
	},

	move: function(access_token,refresh_token,fileId,destinationFileId){
		return this.provider.move(access_token,refresh_token,fileId,destinationFileId);
	},

	copy: function(access_token,refresh_token,fileId,destinationFileId){
		return this.provider.copy(access_token,refresh_token,fileId,destinationFileId);
	},

	delete: function(access_token,refresh_token,fileId){
		return this.provider.delete(access_token,refresh_token,fileId);
	},
};


// stream

Service.prototype.stream = {
	createAndUploadStream: function(access_token,refresh_token,destinationFileId,filename){
		return this.provider.createAndUploadStream(access_token,refresh_token,destinationFileId,filename);
	},

	uploadStream: function(access_token,refresh_token,fileId,overwrite,parent_rev){
		return this.provider.uploadStream(access_token,refresh_token,fileId,overwrite,parent_rev);
	},

	downloadStream: function(access_token,refresh_token,fileId){
		return this.provider.downloadStream(access_token,refresh_token,fileId);
	},
};


// share

Service.prototype.share = {
	list: function(access_token,refresh_token,fileId){
		return this.provider.list(access_token,refresh_token,fileId);
	},

	get: function(access_token,refresh_token,fileId){
		return this.provider.get(access_token,refresh_token,fileId);
	},

	new: function(access_token,refresh_token,fileId,   others){
		return this.provider.new(access_token,refresh_token,fileId,   others);
	},

	update: function(access_token,refresh_token,fileId,   others){
		return this.provider.update(access_token,refresh_token,fileId,   others);
	},

	remove: function(access_token,refresh_token,fileId){
		return this.provider.remove(access_token,refresh_token,fileId);
	},
};

Service.client = function(access_token,refresh_token){
	return new Client(this,access_token,refresh_token);
};

module.exports = Service;
