// polytrix-core-api/drives/googledrive/auth/index.js
var config = require('../../config').googledrive;
var filefog = require('../filefog');
var Q = require('q');


module.exports.authUrl = function(){
	return 'https://accounts.google.com/o/oauth2/auth?' +
	'approval_prompt=force&access_type=offline&scope=' +
	'https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email%20' +
	config.scope +
	'&response_type=code&' +
	'client_id=' + config.client_key +
	'&redirect_uri=' + config.redirect_url;
};

module.exports.getToken = function(code){
	return filefog.provider.oAuthGetAccessToken(code)
	.then(function(results)
	{
		console.log(results);
		return {
			expires_on: new Date(results.expires_on),
			access_token: results.access_token,
			refresh_token: results.refresh_token
		};
	})
	.catch(function(err){
		console.log(err);
		return err;
	});
};

module.exports.renewToken = function(refresh_token){
	return filefog.provider.oAuthRefreshAccessToken({refresh_token:refresh_token})
	.then(function(newOAuthData){
		return newOAuthData;
	})
	.catch(function(err){
		console.log('core api googledrive renewToken() error');
		console.log(err);
		return err;
	});
};
