// polytrix-core-api/drives/onedrive/operation/index.js
var Q = require('q');
var authenticator = require('../authenticator');

// operation
//	create(access_token,refresh_token,destinationFileId,is_folder,filename)
//	rename(access_token,refersh_token,fileId,newName)
//	move(access_token,refresh_token,fileId,destinationFileId)
//	copy(access_token,refresh_token,fileId,destinationFileId)
//	delete(access_token,refresh_token,fileId)

module.exports.create = function(access_token,refresh_token,destinationFileId,is_folder,filename){
	if(!is_folder){
		return createBlankFile(access_token,destinationFileId,filename);
	}else{
		return createFolder(access_token,destinationFileId,filename);
	}
};

function createBlankFile(access_token,destinationFileId,filename){
	var url='https://apis.live.net/v5.0/'+destinationFileId+'/files/' + filename;
	return authenticator.put(access_token,url,{qs:{overwrite:'ChooseNewName'},expectedStatus:201})
	.then(function(body){
		return transform(JSON.parse(body));
	})
	.catch(function(err){
		return transfromError(err);
	});
}
// {"source": "https://bv7pda.bn1301.livefilestore.com/y2mAJpmUmCW445s3kyX8yUWWnpsGtikH0PfI6xrbLYKv54AshbnkuP8GgebygEyif79YPXAwGQ-a_9bV-NIbQtGLjFxXApcRZyMzqfZj4K}wQgFIs6xYsBgNyoadFzS0bZxtwkk4GTTaZcNpDYR9RoN-A/testing.txt?download&psid=1"}

function createFolder(access_token,destinationFileId,foldername){
	var url = 'https://apis.live.net/v5.0/' + destinationFileId;
	return authenticator.post(access_token,url,{expectedStatus:201,
		json:{
			name:foldername
		}
	},true)
	.then(function(body){
		return transform(body);
	})
	.catch(function(err){
		return transfromError(err);
	});
}

module.exports.rename = function(access_token,refresh_token,fileId,newName){
	var url = 'https://apis.live.net/v5.0/' + fileId;
	return authenticator.put(access_token,url,{
		json:{
			name: newName
		}
	},true)
	.then(function(body){
		return transform(body);
	})
	.catch(function(err){
		return transfromError(err);
	});
};

module.exports.move = function(access_token,refresh_token,fileId,destinationFileId){
	var url = 'https://apis.live.net/v5.0/' + fileId;
	return authenticator('MOVE',access_token,url,{
		json:{
			destination: destinationFileId
		},
		expectedStatus: 201
	},true)
	.then(function(body){
		return transform(body);
	})
	.catch(function(err){
		return transfromError(err);
	});

};

module.exports.copy = function(access_token,refresh_token,fileId,destinationFileId){
	var url = 'https://apis.live.net/v5.0/' + fileId;
	return authenticator('COPY',access_token,url,{
		json:{
			destination: destinationFileId
		},
		expectedStatus: 201
	},true)
	.then(function(body){
		return transform(body);
	})
	.catch(function(err){
		return transfromError(err);
	});
};

module.exports.delete = function(access_token,refresh_token,fileId){
	var url = 'https://apis.live.net/v5.0/' + fileId;
	return authenticator.delete(access_token,url,{expectedStatus: 204})
	.then(function(response){
		return {success:true};
	})
	.catch(function(err){
		return transfromError(err);
	});
};

function transform(body){
	var data = {
		success:true,
		identifier:body.id,
		name:body.name,
		_raw:body
	};
	return data;
}

function transfromError(err){
	console.log(err);
	var result = {
		success: false,
		error_code: err.status,
		msg: err.code=='resource_not_found' ? '404 file not found' : err.message
	};
	return result;
}
