// polytrix-core-api/drives/tesseract/auth/index.js
var config = require('../../config').tesseract;
var request = require('request');
var Q = require('q');



module.exports.createUser = function(username){
	var options = {
		url: config.domain + '/user/create/',
		method : 'GET',
		qs: {'username': username},
		headers: {
			'Authorization' : 'API ' + config.apiKey
		}
	};
	return request(options);
};

module.exports.authUrl = function(){
};

module.exports.getToken = function(userid){
	var options = {
		qs: {'userId': userid, 'scope': 'all'}
	};
	return authenticator('GET',config.apiKey,config.domain + '/user/auth/',options);
};

module.exports.renewToken = function(refresh_token){
	var options = {
		qs: {'refresh_token': refresh_token}
	};
	return authenticator('POST', config.apiKey, config.domain + '/user/auth/', options)
};

var authenticator = function(method,api_key,api_url,options){
	options = options || {};
	options.url = api_url;
	options.expectedStatus = options.expectedStatus || 200;
	options.headers = options.headers || {};
	options.headers.Authorization = "API " + api_key;
    options.rejectUnauthorized = false;
	
	return Q.Promise(function(resolve,reject,notify){
		options.method = method.toUpperCase();
		request(
			options,
			function(error,response,body){
				if(!error){
					if(response.statusCode == options.expectedStatus){
                        if(response.headers['content-type'].search('application/json') !== -1){
                            body = JSON.parse(body);
                        }
						resolve(body);
					}else{
						console.log('response.statusCode: ' + response.statusCode);
						if(typeof body === 'string'){
							body = JSON.parse(body);
						}
						body.status = response.statusCode;
						console.log(body);
						reject(body);
					}
				}else{
					reject(error);
				}
			}
		);
	});
};
