// polytrix-core-api/drives/googledrive/metadata/index.js
var filefog = require('../filefog');
var Q = require('q');
var Authenticator = require('../authenticator');
// metadata
//	get(access_token,refresh_token,fileId)
//	listFile(access_token,refresh_token,fileId)
//	downloadUrl(access_token,refresh_token,fileId)

module.exports.get = function(access_token,refresh_token,fileId){
	return filefog.client(access_token,refresh_token)
		.then(function (client) {
			return client.getFileInformation(fileId);
		}).then(function(result){
			result.success = true;
			return result;
		}).catch(function (err){
			console.log(err);
			var result = {
				success: false,
				error_code: err.code,
				msg: err.code==404 ? '404 file not found' : err.message
			};
			return result;
		});
};

module.exports.listFile = function(access_token,refresh_token,fileId){
	return filefog.client(access_token,refresh_token)
		.then(function (client) {
			return client.retrieveFolderItems(fileId);
		}).then(function(result){
			result.success = true;
			return result;
		}).catch(function (err){
			console.log(err);
			var result = {
				success: false,
				error_code: err.code,
				msg: err.code==404 ? '404 file not found' : err.message
			};
			return result;
		});
};

module.exports.downloadUrl = function(access_token,refresh_token,fileId){
	var metadata;
	var url = 'https://www.googleapis.com/drive/v2/files/' + fileId;
	
	return Authenticator.get(access_token,url)
	.then(function(body){
		metadata = JSON.parse(body);
		return 'OK';
	})
	.then(function(okString){
		var url = 'https://www.googleapis.com/drive/v2/files/' + metadata.id + '/permissions';
		var json = {
			role: 'reader',
			type: 'anyone',
		};
		return authenticator.post(access_token,refresh_token,url,{json:json});
	})
	.then(function(body){
		return body.id;
	})
	.then(function(permissionId){
		//remove the permission after one second
		setTimeout(function(){
			var url = 'https://www.googleapis.com/drive/v2/files/' + metadata.id + '/permissions/' + permissionId;
			authenticator.delete(access_token,refresh_token,url,{expectedStatus:204})
			.catch(function(err){
				console.log(err);
			})
			.done();
		},1000);

		return metadata.webContentLink;
	});
};


//metadata (folder)
// { is_file: false,
//   is_folder: true,
//   etag: '"zWM2D6PBtLRQKuDNbaQNSNEy5BE/MTM0NjE3MjU1MTU2Nw"',
//   identifier: '0AOc3nh6wFb5CUk9PVA',
//   parent_identifier: '',
//   mimetype: 'application/vnd.google-apps.folder',
//   created_date: Wed Aug 29 2012 00:49:11 GMT+0800 (HKT),
//   modified_date: Wed Aug 29 2012 00:49:11 GMT+0800 (HKT),
//   name: '我的雲端硬碟',
//   description: '',
//   checksum: undefined,
//   file_size: undefined,
//   _raw: 
//    { kind: 'drive#file',
//      id: '0AOc3nh6wFb5CUk9PVA',
//      etag: '"zWM2D6PBtLRQKuDNbaQNSNEy5BE/MTM0NjE3MjU1MTU2Nw"',
//      selfLink: 'https://www.googleapis.com/drive/v2/files/0AOc3nh6wFb5CUk9PVA',
//      alternateLink: 'https://docs.google.com/folderview?id=0AOc3nh6wFb5CUk9PVA&usp=drivesdk',
//      iconLink: 'https://ssl.gstatic.com/docs/doclist/images/icon_11_collection_list.png',
//      title: '我的雲端硬碟',
//      mimeType: 'application/vnd.google-apps.folder',
//      labels: 
//       { starred: false,
//         hidden: false,
//         trashed: false,
//         restricted: false,
//         viewed: false },
//      createdDate: '2012-08-28T16:49:11.567Z',
//      modifiedDate: '2012-08-28T16:49:11.567Z',
//      markedViewedByMeDate: '1970-01-01T00:00:00.000Z',
//      version: '2',
//      parents: [],
//      userPermission: 
//       { kind: 'drive#permission',
//         etag: '"zWM2D6PBtLRQKuDNbaQNSNEy5BE/CBoOnbDs3kAhsf8c7lrjg-dXTHg"',
//         id: 'me',
//         selfLink: 'https://www.googleapis.com/drive/v2/files/0AOc3nh6wFb5CUk9PVA/permissions/me',
//         role: 'owner',
//         type: 'user' },
//      quotaBytesUsed: '0',
//      ownerNames: [ 'Ka Ho So' ],
//      owners: [ [Object] ],
//      lastModifyingUserName: 'Ka Ho So',
//      lastModifyingUser: 
//       { kind: 'drive#user',
//         displayName: 'Ka Ho So',
//         isAuthenticatedUser: true,
//         permissionId: '09966738586574842813',
//         emailAddress: 'ooloos555@gmail.com' },
//      editable: true,
//      copyable: false,
//      writersCanShare: true,
//      shared: false,
//      appDataContents: false },
//   success: true }




// listFile
// { total_items: null,
//   content: 
//    [ { is_file: true,
//        is_folder: false,
//        etag: '"zWM2D6PBtLRQKuDNbaQNSNEy5BE/MTM2Mzg1MjEzNjY4Ng"',
//        identifier: '0B-c3nh6wFb5CcGJFWmVhTzA0RTQ',
//        parent_identifier: '0AOc3nh6wFb5CUk9PVA',
//        mimetype: 'audio/mpeg',
//        created_date: Thu Mar 21 2013 15:48:29 GMT+0800 (HKT),
//        modified_date: Thu Mar 21 2013 15:48:56 GMT+0800 (HKT),
//        name: '12 卷五錄音材料.mp3',
//        description: '',
//        checksum: 'bc0e564001fb92bd36bbd0abb6dcd33f',
//        file_size: '11936101',
//        _raw: [Object] },
//      { is_file: true,
//        is_folder: false,
//        etag: '"zWM2D6PBtLRQKuDNbaQNSNEy5BE/MTM1NjQxMzM3MjUyMA"',
//        identifier: '1-10OjRM_opXPhFe-2ONV3l_SV6w0ZCIarJG3kK7ujI4',
//        parent_identifier: '0AOc3nh6wFb5CUk9PVA',
//        mimetype: 'application/vnd.google-apps.document',
//        created_date: Tue Dec 25 2012 13:21:35 GMT+0800 (HKT),
//        modified_date: Tue Dec 25 2012 13:29:32 GMT+0800 (HKT),
//        name: '無標題文件',
//        description: '',
//        checksum: undefined,
//        file_size: undefined,
//        _raw: [Object] },
//      { is_file: true,
//        is_folder: false,
//        etag: '"zWM2D6PBtLRQKuDNbaQNSNEy5BE/MTM1NTI5ODAzNzkxNQ"',
//        identifier: '0B-c3nh6wFb5Ccm56Ui15Z3RWTlk',
//        parent_identifier: '0AOc3nh6wFb5CUk9PVA',
//        mimetype: 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
//        created_date: Wed Dec 12 2012 15:40:37 GMT+0800 (HKT),
//        modified_date: Wed Dec 12 2012 15:40:37 GMT+0800 (HKT),
//        name: 'SBA.docx',
//        description: '',
//        checksum: '2adb87ddf9b03d495cb055dc9418bb1a',
//        file_size: '35282',
//        _raw: [Object] },
//      { is_file: true,
//        is_folder: false,
//        etag: '"zWM2D6PBtLRQKuDNbaQNSNEy5BE/MTM1MzkyMjQwNzU0Mw"',
//        identifier: '0B-c3nh6wFb5CNUtXbWdXRE1wM2M',
//        parent_identifier: '0AOc3nh6wFb5CUk9PVA',
//        mimetype: 'application/zip',
//        created_date: Mon Nov 26 2012 17:33:27 GMT+0800 (HKT),
//        modified_date: Mon Nov 26 2012 17:33:27 GMT+0800 (HKT),
//        name: 'new.zip',
//        description: '',
//        checksum: '0e7922c4530f68c721c893498ffbe5ba',
//        file_size: '1241433',
//        _raw: [Object] } ],
//   success: true }
