// polytrix-core-api/drives/googledrive/metadata/index.js
var filefog = require('../filefog');
var Q = require('q');
var Authenticator = require('../authenticator');
// metadata
//	get(access_token,refresh_token,fileId)
//	listFile(access_token,refresh_token,fileId)
//	downloadUrl(access_token,refresh_token,fileId)

module.exports.get = function(access_token,refresh_token,fileId){
	return filefog.client(access_token,refresh_token)
		.then(function (client) {
			return client.getFileInformation(fileId);
		}).then(function(result){
			result.success = true;
			return result;
		}).catch(function (err){
			console.log(err);
			var result = {
				success: false,
				msg: err.code==404 ? '404 file not found' : err.message
			};
			return result;
		});
};

module.exports.listFile = function(access_token,refresh_token,fileId){
	return filefog.client(access_token,refresh_token)
		.then(function (client) {
			return client.retrieveFolderItems(fileId);
		}).then(function(result){
			result.success = true;
			return result;
		}).catch(function (err){
			console.log(err);
			var result = {
				success: false,
				msg: err.code==404 ? '404 file not found' : err.message
			};
			return result;
		});
};

module.exports.downloadUrl = function(access_token,refresh_token,fileId){
	var metadata;
	var url = 'https://www.googleapis.com/drive/v2/files/' + fileId;
	
	return Authenticator.get(access_token,url)
	.then(function(body){
		metadata = JSON.parse(body);
		return 'OK';
	})
	.then(function(okString){
		var url = 'https://www.googleapis.com/drive/v2/files/' + metadata.id + '/permissions';
		var json = {
			role: 'reader',
			type: 'anyone',
		};
		return authenticator.post(access_token,refresh_token,url,{json:json});
	})
	.then(function(body){
		return body.id;
	})
	.then(function(permissionId){
		//remove the permission after one second
		setTimeout(function(){
			var url = 'https://www.googleapis.com/drive/v2/files/' + metadata.id + '/permissions/' + permissionId;
			authenticator.delete(access_token,refresh_token,url,{expectedStatus:204})
			.catch(function(err){
				console.log(err);
			})
			.done();
		},1000);

		return metadata.webContentLink;
	});
};
