// polytrix-core-api/drives/tesseract/metadata/index.js
var authenticator = require('../authenticator');
var Q = require('q');
// metadata
//	get(access_token,refresh_token,fileId)
//	listFile(access_token,refresh_token,fileId)
//	downloadUrl(access_token,refresh_token,fileId)

module.exports.get = function(access_token,refresh_token,fileId){
	return filefog.client(access_token,refresh_token)
	.then(function (client) {
		return client.getFileInformation(fileId);
	}).then(function(result){
		result.success = true;
		if(result._raw.type == 'album'){
			result.is_file = false;
			result.is_folder = true;
		}
		return result;
	}).catch(function (err){
		transfromError(err);
	});
};

module.exports.listFile = function(access_token,refresh_token,fileId){
	return filefog.client(access_token,refresh_token)
	.then(function (client) {
		return client.retrieveFolderItems(fileId);
	}).then(function(result){
		result.success = true;
		result.content = result.content.map(function(file){
			if(file._raw.type == 'album'){
				file.is_file = false;
				file.is_folder = true;
			}
			return file;
		});
		return result;
	}).catch(function (err){
		transfromError(err);
	});
};

module.exports.downloadUrl = function(access_token,refresh_token,fileId){
	return Authenticator.get(access_token,'https://apis.live.net/v5.0/'+fileId+'/content?suppress_redirects=true&download=true')
	.then(function(body){
		body = JSON.parse(body);
		return body.location;
	});
};

function transfromError(err){
	console.log(err);
	var result = {
		success: false,
		error_code: err.status,
		msg: err.code=='resource_not_found' ? '404 file not found' : err.message
	};
	return result;
}
