// polytrix-core-api/drives/tesseract/metadata/index.js
var authenticator = require('../authenticator');
var config = require('../../config').tesseract;
var Q = require('q');
// metadata
//	get(access_token,refresh_token,fileId)
//	listFile(access_token,refresh_token,fileId)
//	downloadUrl(access_token,refresh_token,fileId)

module.exports.get = function(access_token,refresh_token,fileId){
    var url = config.domain + '/vault/file/' + fileId + '/metadata';
    return authenticator.get(access_token,url)
        .then(transformMetadata)
        .then(function(result){
            result.success = true;
            return result;
        }).catch(transformError);
};

module.exports.listFile = function(access_token,refresh_token,fileId){
    var url = config.domain + '/vault/file/' + fileId + '/list';
    return authenticator.get(access_token,url)
        .then(transformListFile)
        .then(function(result){
            result.success = true;
            return result;
        }).catch(transformError);
};

module.exports.downloadUrl = function(access_token,refresh_token,fileId){
	throw new Error('should not call this');
};

function transformMetadata(raw){
    return {
        is_file:!raw.is_folder,
        is_folder:raw.is_folder,
        identifier:raw.identifier,
        parent_identifier:raw.parent_identifier,
        created_date:raw.created_date,
        modified_date:raw.modified_date,
        name:raw.name,
        description:raw.description,
        _raw:raw
    };
}

function transformListFile(raw){
    var arr = raw.map(function(record){
        return transformMetadata(record);
    });
    return {
        content:arr
    };
}

function transformError(err){
	console.log(err);
	var result = {
		success: false,
		error_code: err.status,
		msg: err.code=='resource_not_found' ? '404 file not found' : err.message
	};
	return result;
}
